### ISSUE TITLE
Double display output in GRUB in QEMU v3.1

### ISSUE THREAD

**Issue by ehlers** [GNS3/gns3-server#1516]
GNS3 v2.1.12 in Debian 10 (buster) with QEMU v3.1

Just upgraded my own GNS3 VM from Debian stretch to buster. In general the QEMU v3.1 from buster works much better than the QEMU v2.8 from stretch, the issues documented in  <https://www.bernhard-ehlers.de/blog/2017/08/11/gns3-qemu-issues.html> are gone.

But I got a small new issue: While the QEMU VM uses the BIOS for I/O, for example while in the GRUB boot loader, each character is output twice. This happens only, when the QEMU option "-nographic" is used, so all VMs with telnet output are affected. Using the option "-display none" instead works better. Then each character is output only once.

My current solution adds "-display none" to the QEMU options, when no other display related option is set. I tested this only with QEMU v3.1, I don't know if this is working with older QEMU. At least the "ancient" version 0.14 doesn't support the "-display" option.

```
diff --git a/gns3server/compute/qemu/qemu_vm.py b/gns3server/compute/qemu/qemu_vm.py
index 16c9ebdf..43254bb6 100644
--- a/gns3server/compute/qemu/qemu_vm.py
+++ b/gns3server/compute/qemu/qemu_vm.py
@@ -1610,8 +1610,8 @@ class QemuVM(BaseNode):
             return []
         if len(os.environ.get("DISPLAY", "")) > 0:
             return []
-        if "-nographic" not in self._options:
-            return ["-nographic"]
+        if not any(opt in self._options for opt in ["-display", "-nographic", "-curses", "-sdl" "-spice", "-vnc"]):
+            return ["-display", "none"]
         return []
 
     def _run_with_kvm(self, qemu_path, options):
```

The impact of this issue is quite small, only very new QEMU versions are affected (for example v3.1) and only while the VM is in the boot loader. So I won't give that a high priority.

**Comment by ehlers** [comment]
Just verified, that this issue with double output in BIOS doesn't happen with QEMU v2.5 in the official GNS3VM and doesn't happen with QEMU 2.8 in Debian 9 (Stretch). So I assume it's a QEMU v3.x issue.

Just thought about the best way to fix it. My current solution follows the following design:
The disabling of the graphics (-nographics / -display none) is solely handled in the gns3-server.

The code in the gns3-gui (gns3/modules/qemu/dialogs/qemu_vm_wizard.py and gns3/registry/config.py), which adds "-nographics / -vga none -vnc none) should be deleted.

The gns3-server modifications:

```
diff --git a/gns3server/compute/qemu/qemu_vm.py b/gns3server/compute/qemu/qemu_vm.py
index 16c9ebdf..5f183668 100644
--- a/gns3server/compute/qemu/qemu_vm.py
+++ b/gns3server/compute/qemu/qemu_vm.py
@@ -1601,17 +1601,18 @@ class QemuVM(BaseNode):
 
         return network_options
 
-    def _graphic(self):
+    @asyncio.coroutine
+    def _disable_graphics(self):
         """
-        Adds the correct graphic options depending of the OS
+        Disable graphics depending of the QEMU version
         """
 
-        if sys.platform.startswith("win"):
-            return []
-        if len(os.environ.get("DISPLAY", "")) > 0:
-            return []
-        if "-nographic" not in self._options:
-            return ["-nographic"]
+        if not any(opt in self._options for opt in ["-display", "-nographic", "-curses", "-sdl" "-spice", "-vnc"]):
+            version = yield from self.manager.get_qemu_version(self.qemu_path)
+            if version and parse_version(version) >= parse_version("3.0"):
+                return ["-display", "none"]
+            else:
+                return ["-nographic"]
         return []
 
     def _run_with_kvm(self, qemu_path, options):
@@ -1678,7 +1679,8 @@ class QemuVM(BaseNode):
             raise QemuError("Console type {} is unknown".format(self._console_type))
         command.extend(self._monitor_options())
         command.extend((yield from self._network_options()))
-        command.extend(self._graphic())
+        if self._console_type == "telnet":
+            command.extend((yield from self._disable_graphics()))
         if additional_options:
             try:
                 command.extend(shlex.split(additional_options))
```

I assume, that the change happens with QEMU v3.0, that has to be verified. Ubuntu 18.04 with QEMU v2.11 is a good candidate for further tests.

I successfully tested that with the GNS3VM (QEMU 2.5), Debian 9 (QEMU 2.8), Debian 10 (QEMU 3.1) and OS X (QEMU 0.14). I haven't tested with Windows.

I'm unsure about QEMU 0.11. I don't know, if that supports "-nographic". Maybe it needs "-vga none -vnc none". But unless it's still used in Windows, I suggest to no longer support that version.

**Comment by ehlers** [comment]
Just saw, that you committed my patch. As my patch changed the QEMU command line, the tests are failing now. So you have to fix them. Furthermore I have no idea, what happens on Windows, have you tested that?

**Comment by grossmj** [comment]
I am fixing the tests. Regarding Windows, still need to investigate what is the problem, looks like that for some reasons cx_freeze cannot find PyQt5.Core anymore.

**Comment by grossmj** [comment]
My bad, I thought you were talking about another Windows problem we have at the moment. I will test on Windows with the next nightly build.

**Comment by ehlers** [comment]
I've got a Windows 2008 64-bit VM and I installed the nightly build GNS3-2.1.2019-02-18-all-in-one in it. With the QEMU 2.4.0 it works, but during the start it took some time until the communication to the gns3server was completed. No idea if that's normal.

QEMU v0.11 didn't work, here the communication to the gns3server stalls. After some time the GUI tells me, that no communication to the gns3server is possible. But as other actions before and afterwards have no issue communication with the server, I assume the server hangs somewhere during the start of the qemu process.

**Comment by ehlers** [comment]
Re-tested with the official GNS3 v2.1.12. It's behaving the same, so at least it didn't get worse.

QEMU 2.4.0 works, but needs "forever" to start (several seconds).

The QEMU 0.11 VM doesn't start, error message:
Could not start QEMU console [WinError 1225] The remote computer refused the network connection

Starting QEMU 0.11 with that VM image by CMD works.

**Comment by ehlers** [comment]
Found the issue with QEMU v0.11: It gets started with "-smp cpus=1". But that's invalid for v0.11, that version needs "-smp 1". So either some poor soul (@grossmj) has to fix that, or drop qemu 0.11.

**Comment by grossmj** [comment]
Thanks for the tests. I think we should officially drop Qemu v0.11 in GNS3 2.2.

**Comment by grossmj** [comment]
I will also upgrade QEMU 2.4.0 to version 3.1.0 (the one bundled in our Windows installer). I prefer to do this in GNS3 2.2 as well. Do you think there is anything left to change in our 2.1 branch?

**Comment by ehlers** [comment]
For v2.1 I'm fine, for V2.2 I have two open aspects:
- The idea of my changes are, that the GUI doesn't have to modify the QEMU options for disabling the graphics, all should be handled by the gns3-server. So you might remove the corresponding code in gns3-gui (gns3/modules/qemu/dialogs/qemu_vm_wizard.py and gns3/registry/config.py).
- What QEMU version for OS X? v0.14 is quite dated as well.

**Comment by grossmj** [comment]
> The idea of my changes are, that the GUI doesn't have to modify the QEMU options for disabling the graphics, all should be handled by the gns3-server. So you might remove the corresponding code in gns3-gui (gns3/modules/qemu/dialogs/qemu_vm_wizard.py and gns3/registry/config.py).

I agree, I will change this.

> What QEMU version for OS X? v0.14 is quite dated as well.

We should try to update this however I remember that compiling and embedding Qemu on macOS is not an easy task. Do you know if someone provide Qemu binaries for macOS? (just like this person does for Windows: https://qemu.weilnetz.de/w64/).

**Comment by ehlers** [comment]
> Do you know if someone provide Qemu binaries for macOS?

The official way is to install it by Homebrew or MacPorts <https://www.qemu.org/download/#macos>

Alternatively you can use the slightly outdated v2.10.1 binaries from <https://wiki.qemu.org/Hosts/Mac>. I just integrated that into my /Applications/GNS3.app and my basic tests (telnet and VNC console) were successful.

What I did:
- removed the old qemu in /Applications/GNS3.app/Contents/Resources/
- Unpacked QEMU 2.10.1.zip and moved the QEMU folder to /Applications/GNS3.app/Contents/Resources/qemu
- Renamed /Applications/GNS3.app/Contents/Resources/qemu/Programs to bin

**Comment by grossmj** [comment]
PR opened: https://github.com/GNS3/gns3-gui/pull/2719

Also, for GNS3 2.2:
* Qemu 2.4.0 has been replaced by version 3.1.0 in Windows installer.
* Qemu 0.14.0 has been replaced by version 2.10.1 in macOS app.
* Qemu 0.11.0 has been removed from Windows installer.

**Comment by grossmj** [comment]
Everything should be in order now. Let's close this.
